### DISCUSSION TITLE
A year's worth of suggestions

### DISCUSSION THREAD

**Discussion by SiIentNinja** [ScratchAddons/ScratchAddons#7905] [Ideas]
### Hi,

I've been collecting ideas over the past year, adding them to the list whenever I feel an inconveniency with Scratch. Hopefully I can provide some insight from a different perspective, describing some of the things that personally frustrate me about Scratch, maybe many other people as well.

I have to say, when I was going through the list before I posted this, I did find quite a lot of suggestions that I had noted down that got added recently, I am really happy about that, and grateful to all the people who worked hard to create these addons.

I have tried to keep this post as concise as possible and removed a lot of suggestions, I hope it's not too long
Some suggestions are ok, some are really dumb, but hopefully I can be of help in some way.

**Starting off, I would like to briefly mention a few of the suggestions from last year** (just to refresh your mind)
 - fps counter + project timer at top (like the clone counter)- useful for debugging 
 -  Show hidden options (similar to point towards random direction addon)
        - show if touching self option (for clone detection)
        - show previous costume option
        - show last and random options in item _ of list (when right clicked)
        - show "say" in the backdrop
 - Search bar in user's projects/studios/favourites/loves/following/followers in profile (the full page lists of these) 
 - more fonts in art editor- show more fonts that work in scratch ("scratch" font and https://scratch.mit.edu/projects/667310164/) 
___________________________________________________________________________________________________________________________________

So yeah, here's the list of suggestions. Set aside some time when you are available to sit back and read through this thing.  I have categorised them for your convenience

**Code editor addons**

Comments
 - sometimes when you press undo a random comment disappears (even if it wasn't the last action), and when you redo it the text in the comment is gone (some way to fix this?)
 - detach comment from block
 - when drag scripts with comments into another sprite, make the comment not get deleted
 - when create a comment make it attach to the left side of the block (rather than the right), so if you have a lot of them you can line them neatly in a line on the left (rather than its location changing based on the width of the block it is attached to)

Custom blocks
 - drag custom block parameters from palette (so if define is really long you don't have to scroll back up)
 - Make custom blocks show a "greyed out" name of the parameters (that gets removed when you type something in it)
 - When dragging a script with a custom block in it to another sprite, make its definition also get copied 

Variables/lists
 - allow importing and exporting of lists without having to show the list
 - rearrangeable variables in the menu instead of alphabetical order, can add categories
 - right click a variable/list in the menu to lock it- like for example if you have a lookup table or rom you can't accidentally add/change something or clear it
 - allow typing in letters (rather than just numbers) in list blocks so can easily type in "last" or "random"
 - make restore costume stay after you reload (maybe store in backpack)
 
Misc
 - in the _ of _ sensing block, when changing between sprites the first "_" doesn't turn into x position and instead remains at what it was at before
 - multiple "tabs" of code in one sprite, allow categorising the code into different tabs
 - the grab single block works with nested operators (e.g sin or not), pulling out just that operator and not the things in it (if the operator has a single input).
 - control drag or something to multi select scripts in editor (for moving/deleting/copying them)

**Costume editor addons**

 - stop "when _ key pressed" registering when arrow keys are pressed to move elements in the art editor
 - split a line in costume editor (click on reshape dots)
 - a shortcut to make resizing in the costume editor to also shrink the line and outline thickness to match the shrinking (instead of it staying the same)
 - in costume editor there can be a "select" option but can select by drawing an irregular shape

**Miscellaneous editor addons**

Cloud vars
 - cloud variables don't get turned to non cloud variables when new project is created
 - auto "renaming" the cloud variable each time you save the project to reduce chance of it not working (https://scratch.mit.edu/discuss/topic/12949/?page=101)
 
Backpack
 - rearrange items in backpack (dragging) (eg so can rearrange according to how often you use), categorise backpack- eg custom blocks, costumes, sprites etc (or filter button)
 
Other
 - when first create a project, instead of the import custom project addon, you can select from a list of templates (that you can add to)
 - when create sprite/costume, the name is automatically selected and highlighted
 - setting that makes the blocks not under a hat block not count towards the block count

**Project page addons**

 - Search bar in add projects to studio in the project page (if you have a lot of studios and you are looking to add to a specific one)
 - When press tab hangul filler character "ㅤ" is typed in the project/studio description (because if you use space characters multiple next to each other won't show for others)
 - If in trending say #_ in trending (max 100 or something)
 - Comment count on project page
 - in the "add to studio" popup in the project page, put studios in which you cannot re-add the project (e.g propose projects to be featured) into a separate section, or add confirmation when removing the project
 - for especially long remix chains, allow further scrolling of the remix tree beyond max (eg the third page of https://scratch.mit.edu/projects/71656786/remixtree/, it curves off the bottom of the screen)

**Studio addons**

 - Search bar in add projects to studio popup screen (in studio) if you have a lot of projects you can just search it up instead of scrolling through all.
 - search bar to search projects in studio
 - Exact count in studio projects, creators etc (rather than 100+) (on the studio screen tabs), also red if the manager count is nearing max

**"My stuff" addons**

 - hover over heart and faves symbols to see percentage of viewers who did this (in my stuff)
 - In my stuff in studios if you are a manager say "manager" instead of "curator"

**Miscellaneous addons**

 - If a person you follow both loves and faves a project, it will show up as one message instead of 2 on what's happening, to make more room
 - The SDS news notification immediately updates when the new SDS is created rather than waiting for a few days
 - advanced search (helps find a project you don't know the name) (e.g contains key words in description, time shared etc)
 - "screen time", showing how long spent creating projects & how long spent exploring other people's projects, forums etc
 - in filter messages addon- setting to make the removed messages don't go to the message count, and also not make them show in the "scratch messaging" tab addon

And lastly, of course- turn move_steps into move_step if it is 1 (also degrees) lol
___________________________________

I'm not pushing you to add any of these, as previously said, I hope to provide some insight on my personal frustrations with scratch and some suggestions to help tackle these.

Perhaps I've inspired you to add some similar things to my suggestions, or maybe by reading this I've revealed to you with a new low of boredom and mental fatigue. But I have tried my best to collect lots of (hopefully useful) suggestions throughout the year, I really hope my post has helped you in some way.

Thank you all for your hard work, Scratch addons has played a significant role in my experience with Scratch, I wouldn't have reached the level of scratch I am in today without it.

_(ps don't worry this is likely to be the last annual idea dump in a while)
(I miight maybe make another post just on my favourites so I can expand on them a bit more, if requested)_

edit: I removed the ones that Samq64 said were bad or already being worked on

**Comment by Samq64** [comment]
Your suggestions are more likely to be noticed if you create a separate issue for each one that hasn't already been suggested and thumbs up or comment on the ones that have.

> fps counter + project timer at top (like the clone counter)- useful for debugging

#3656 adds the FPS counter. I don't think the timer has been suggested before since its monitor can already be shown.

> Show hidden options (similar to point towards random direction addon)

- #6761

> more fonts in art editor- show more fonts that work in scratch ("scratch" font and https://scratch.mit.edu/projects/667310164/)

This was previously rejected (#1646). TurboWarp has it but there are compatibility issues with Scratch.

> Comments

All of those could be part of the better editor comments addon.

> Make custom blocks show a "greyed out" name of the parameters (that gets removed when you type something in it)

#7493 suggests a tooltip instead but this is better.

> add "is compiled?" to palette

- #3848

> rearrangeable variables in the menu instead of alphabetical order, can add categories

Related: #7726

> hover over a variable in the sidebar to see its amount of uses (useful for knowing if there is an unused variable so you can delete it (eg when extracting a sprite from a project (useful for converting projects)), or see which variables are the most used than others etc). Alternatively, do a variable deletion confirmation like "are you sure you want to delete this var, it has _ uses"

Isn't there already a deletion confirmation the number of uses?

> right click a script has a option to copy scratchblock version

- #2336

> if type something like "///" before a value eg when setting a variable, the sb3 file will be edited to make that value invisible (for hiding things like secret codes)

I'm not sure how that would work and don't think it's a good idea.

> collapse nested blocks in control blocks

- #1199

> rearrange items in backpack (dragging), categorise backpack- eg custom blocks, costumes, sprites etc (or filter button)

#7435 suggests adding a search bar.

> import and export backpack (for transferring accounts)

#6161 suggests a backpack that is tied to the browser instead of the account.

> show the layering of every sprite in the debugger

#3755 suggests a layer manager but not in the debugger

> When press tab a special blank space character "ㅤ" is typed in the project/studio description (because if you use space characters multiple next to each other won't show for others)

This seems to niche to be an addon and breaks keyboard navigation.

> Comment count on project page

Since it shows on my stuff I assume there's an API for it.

> If there is a turbowarp link to the project in the project's description, and it has settings on it the turbowarp button addon will also have those settings

- #5707
- #7004

> see if a project is nfe

- #6484

> setting to make the comments statistics on projects in my stuff not include your own comments

I don't think that's possible without making a ton of requests.

> in the "add to studio" popup in the project page, put studios in which you cannot re-add the project (e.g propose projects to be featured) into a separate section, or add confirmation when removing the project

I thought confirm actions already had the option for addon/removing projects from studios but I seem to  be wrong.

> for especially long remix chains, allow further scrolling of the remix tree beyond max (eg the third page of https://scratch.mit.edu/projects/71656786/remixtree/, it curves off the bottom of the screen)

I guess that could be fixed but the addon only links to an existing Scratch page it doesn't modify it.

> can rearrange projects in studios by dragging them around (automatic remove and re-add, only works if manager/host)

That would send too many requests, especially with large studios.

> "folders" addon but for projects in my stuff menu

- #7608

> show how much storage a project takes in the my stuff screen under the last modified (red if it is nearing the limit)

#6507 adds size info in the editor.

> hover over heart and faves symbols to see percentage of viewers who did this

#6919 Adds this but for project pages.

> get notified if someone makes a post in a topic you are currently typing in

This would probably be part of a live commenting addon but we're hesitant to add that.

> Copy & paste scripts from forums and scratch wiki into project editor/add to backpack button

- #2336
- #3225

> extension popup tab with your favourited/saved projects(for easy access)

- #4057

**Reply by SiIentNinja** [reply]
Oh wow. I did not realise there were so many already being worked on. I've been searching in the actual extension as well as the discussions for duplicates, I did not know what an "issue" is. I'll edit the post now to remove those, thanks. But seeing that all these are going to be added really makes me happy.

(I could not create a separate issue for each because that would take too long)

‎

> fps counter + project timer at top (like the clone counter)- useful for debugging
>https://github.com/ScratchAddons/ScratchAddons/pull/3656 adds the FPS counter. I don't think the timer has been suggested before since its monitor can already be shown.

> DNin01: I like the timer display idea. Means you don't have to go into the editor and toggle the timer monitor if you need it.

>add "is compiled?" to palette
>https://github.com/ScratchAddons/ScratchAddons/issues/3848

Yeah last time I requested the for_ in_ but they said no. But I don't think is compiled is a "hidden block" because you don't have to change the inner code to get it. But it's fine, I'll delete this one, it's not that important

>>hover over a variable in the sidebar to see its amount of uses (useful for knowing if there is an unused variable so you can delete it >(eg when extracting a sprite from a project (useful for converting projects)), or see which variables are the most used than others >etc). Alternatively, do a variable deletion confirmation like "are you sure you want to delete this var, it has _ uses"
>
>Isn't there already a deletion confirmation the number of uses?

oh there is my bad, I'm guessing I was confused seeing that it only deletes the variables in the sprite you are currently in. However I think for transferring things between projects it still would be useful to see the amount of uses of each variable without having to attempt to delete each one. (the first bit of the suggestion)

>>if type something like "///" before a value eg when setting a variable, the sb3 file will be edited to make that value invisible (for >>hiding things like secret codes)
>
>I'm not sure how that would work and don't think it's a good idea.

I thought so, however I think it is useful to fix the issue this is trying to address. Is there any other method of doing this?

>>rearrange items in backpack (dragging), categorise backpack- eg custom blocks, costumes, sprites etc (or filter button)
>
>https://github.com/ScratchAddons/ScratchAddons/issues/7435 suggests adding a search bar.

Personally what I had hoped for was like a "library" of each individual category, eg if you want a specific custom block you go to the custom blocks section/tab and go through there
Searching is nice, but personally I think sometimes it can be annoying to use, you need to remember the name of the thing you want. Maybe a filter button, this would be in between

>https://github.com/ScratchAddons/ScratchAddons/issues/6161 suggests a backpack that is tied to the browser instead of the account.

Yess this is much better, I didn't think it was possible


>>When press tab a special blank space character "ㅤ" is typed in the project/studio description (because if you use space characters multiple next to each other won't show for others)
>
>This seems to niche to be an addon and breaks keyboard navigation.

Not niche at all. I cannot count the number of times I find myself going to my clipboard just for this character. For someone who wants their descriptions to look nice and formatted this character is essential (because spaces don't work), I would argue this is more important than using tab to switch between input boxes.

>>in the "add to studio" popup in the project page, put studios in which you cannot re-add the project (e.g propose projects to be featured) into a separate section, or add confirmation when removing the project
>
>I thought confirm actions already had the option for addon/removing projects from studios but I seem to be wrong.

I'm hoping to make this feature only for the studios in which you aren't a curator and "anyone can add projects" is off. For example if someone is trying to advertise by double clicking the list of studios in the add to studio popup, you won't accidentally remove it from something like the sds and not be able to add it back

**Reply by Samq64** [reply]
> Oh wow. I did not realise there were so many already being worked on.

Most new addon issues aren't currently being worked on, they're just there in case someone wants to start working on them in the future.

> Not niche at all. I cannot count the number of times I find myself going to my clipboard just for this character. For someone who wants their descriptions to look nice and formatted this character is essential (because spaces don't work), I would argue this is more important than using tab to switch between input boxes.

Now I see how that could be useful. I was thinking of single line textboxes.
